### PR TITLE
Second order sdc

### PR DESCRIPTION
**Pull request by ikrom96git**
Finally, I have made all of the code available for the second order SDC (Spectral Deferred Correction) paper. Here's what you can find:

1. In the projects' Second_orderSDC folder, I have created a directory where you can access all of the codes needed to implement the plots mentioned in the paper. For more information, please refer to the README file.

2. Additionally, I have included tests for the codes. These tests evaluate the convergence order of the Runge-Kutta-Nystrom method, the Velocity-Verlet scheme, and the SDC method. However, the SDC test only measures local convergence order as it requires less time to run.

3. I want to express my special thanks to @tlunet and @brownbaerchen for providing guidance and assistance in writing the codes.

I hope this clears up the information you provided. Let me know if you have any further questions!

### CODE REVIEW

**Comment by pancetta** [comment]
Nice, thanks a lot! Could you look at the [results](https://github.com/Parallel-in-Time/pySDC/pull/330/checks?check_run_id=15075627397) of codecov and increase the test coverage? I guess `dampedharmonic_oscillator_run_stability.py` and most of the penningtrap stuff could use some testing. If the individual runs are too long, try to create a smaller/shorter test case which goes through the same algorithms as your actual simulations.

File: pySDC/tests/test_projects/test_second_orderSDC/test_convergence.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        expected_order['velocity'] = np.vstack((expected_order['velocity'], string_to_array(items[2])))
+        num_order['position'] = np.vstack((num_order['position'], np.round(string_to_array(items[1]))))
+        num_order['velocity'] = np.vstack((num_order['velocity'], np.round(string_to_array(items[3][:-1]))))
+
+    return num_order, expected_order
+
+
+@pytest.mark.base
+@pytest.mark.parametrize('value', ['position', 'velocity'])
+def test_BorisSDC_horizontal_axis(value):
```
Comment (brownbaerchen) [inline pySDC/tests/test_projects/test_second_orderSDC/test_convergence.py:96]: Do you generate the same results twice and then check the different values? Maybe I misunderstand this, but if this is the case, then best generate the results only once and these both position and velocity at the same time. Then the tests get faster.

File: pySDC/tests/test_projects/test_second_orderSDC/test_convergence.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+    expected_order, num_order = sort_order()
+
+    assert np.isclose(
+        num_order[value][0, :], expected_order[value][0, :], atol=2.6e-1
+    ).all(), 'Expected order in {} {}, got {}!'.format(value, expected_order[value][0, :], num_order[value][0, :])
+
+
+@pytest.mark.base
+@pytest.mark.parametrize('value', ['position', 'velocity'])
```
Comment (brownbaerchen) [inline pySDC/tests/test_projects/test_second_orderSDC/test_convergence.py:116]: Same here

File: pySDC/tests/test_projects/test_second_orderSDC/test_convergence.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    level_params = dict()
+    level_params['restol'] = 1e-16
+    level_params['dt'] = 0.015625
+
+    # initialize sweeper parameters
+    sweeper_params = dict()
+    sweeper_params['quad_type'] = 'GAUSS'
+    sweeper_params['num_nodes'] = 5
+    sweeper_params['do_coll_update'] = True
+    sweeper_params['initial_guess'] = 'zero'  # 'zero', 'spread'
```
Comment (brownbaerchen) [inline pySDC/tests/test_projects/test_second_orderSDC/test_convergence.py:36]: Why do you choose zero here? I found that usually spread works much better.

**Review by brownbaerchen — COMMENTED** [review]
I didn't look through the scripts where you generate the plots, I just looked through the RK sweeper and the tests. I just marked some small things that I think may cause bugs later on when other people implement things on top.

**Comment by ikrom96git** [comment]
> I didn't look through the scripts where you generate the plots, I just looked through the RK sweeper and the tests. I just marked some small things that I think may cause bugs later on when other people implement things on top.

For the test, I haven't implemented the convergence plots for the tests. But, one gets the plots in the project itself.

File: pySDC/tests/test_projects/test_second_orderSDC/test_convergence.py (hunk 1 above)
Comment (ikrom96git) [reply pySDC/tests/test_projects/test_second_orderSDC/test_convergence.py:96]: Thanks for the comments, I will fix these issues right away.

File: pySDC/tests/test_projects/test_second_orderSDC/test_convergence.py (hunk 3 above)
Comment (ikrom96git) [reply pySDC/tests/test_projects/test_second_orderSDC/test_convergence.py:36]: That is right, spread works better but to reach minimal convergence order like in the project. I needed to use zero.

**Comment by pancetta** [comment]
Looks better, but test coverage is still not great, see [comment above](https://github.com/Parallel-in-Time/pySDC/pull/330#issuecomment-1637450826).

**Comment by lisawim** [comment]
I guess that you are familiar with the penning trap problem. Can you write some documentation for the `penningtrap` problem class as mentioned in #286?

**Comment by ikrom96git** [comment]
> I guess that you are familiar with the penning trap problem. Can you write some documentation for the `penningtrap` problem class as mentioned in #286?

Sure, I will write down documentation for the Penning Trap problem.

**Comment by pancetta** [comment]
You made it! Very good!

**Pull request merged**